### PR TITLE
Upgrade controller runtime to v0.17.5 and k8s api to v0.29.5

### PR DESCRIPTION
**Pull request by xliuqq**
### Ⅰ. Describe what this PR does
Upgrade controller runtime to v0.17.5 and k8s api to v0.29.5.

### Ⅱ. Does this pull request fix one issue?

fixes #XXXX

### Ⅴ. Special notes for reviews

### CODE REVIEW

**Comment by cheyang** [comment]
@xliuqq Please fix the code conflict. Thanks.

**Comment by xliuqq** [comment]
@cheyang the e2e-test was passed.

File: cmd/efc/app/efc.go [hunk 1]
```diff
@@ -113,13 +114,15 @@ func handle() {
 
 	utils.NewPprofServer(setupLog, pprofAddr, development)
 
+	// the default webserver port is 9443, no need to set.
```
Comment (cheyang) [inline cmd/efc/app/efc.go:117]: How about changing comment to "the default webhook server port is 9443, no need to set." ?

**Review by cheyang — CHANGES_REQUESTED** [review]

File: cmd/dataset/app/dataset.go [hunk 2]
```diff
@@ -121,14 +123,16 @@ func handle() {
...
+		Scheme: scheme,
+		Metrics: metricsserver.Options{
+			BindAddress: metricsAddr,
+		},
 		LeaderElection:          enableLeaderElection,
 		LeaderElectionNamespace: leaderElectionNamespace,
 		LeaderElectionID:        "dataset.data.fluid.io",
-		Port:                    9443,
-		NewCache:                NewCache(scheme),
+		Cache:                   NewCacheOptions(),
```
Comment (cheyang) [inline cmd/dataset/app/dataset.go:135]: Why using `Cache` instead of `NewCache`?

Comment (xliuqq) [reply cmd/dataset/app/dataset.go:135]: Previous used `cache.BuilderWithOptions` function is removed. The `NewCacheFunc` is changed to LOW LEVEL PRIMITIVE! We only need to configure the cache options, no need to re-implement a cache function.

File: cmd/efc/app/efc.go (hunk 1 above)
Comment (xliuqq) [reply cmd/efc/app/efc.go]: fixed.

File: cmd/dataset/app/dataset.go (hunk 2 above)
Comment (cheyang) [reply cmd/dataset/app/dataset.go:135]: Got it. Thanks.

File: cmd/juicefs/app/juicefs.go [hunk 3]
```diff
@@ -121,20 +121,21 @@ func handle() {
...
+		Scheme: scheme,
+		Metrics: metricsserver.Options{
+			BindAddress: metricsAddr,
+		},
 		LeaderElection:          enableLeaderElection,
 		LeaderElectionNamespace: leaderElectionNamespace,
 		LeaderElectionID:        "juicefs.data.fluid.io",
-		Port:                    9443,
-		NewCache:                juicefsctl.NewCache(scheme),
+		Cache:                   juicefsctl.NewCache(),
```
Comment (cheyang) [inline cmd/juicefs/app/juicefs.go:138]: How about using `juicefsctl.NewCacheOption()`?

**Review by cheyang — CHANGES_REQUESTED** [review]

File: cmd/dataset/app/dataset.go [hunk 4]
```diff
@@ -232,21 +236,22 @@ func handle() {
...
-		})}
+		cronJobKey = &batchv1beta1.CronJob{}
 	}
 
-	return cache.BuilderWithOptions(cache.Options{
-		Scheme:            scheme,
-		SelectorsByObject: selectors,
-	})
+	return cache.Options{
+		ByObject: map[client.Object]cache.ByObject{
```
Comment (cheyang) [inline cmd/dataset/app/dataset.go:249]: Do we need to add scheme?

```golang
return cache.Options{
		Scheme: scheme,
		ByObject: map[client.Object]cache.ByObject{
			cronJobKey: {
				Label: labels.SelectorFromSet(labels.Set{
					common.JobPolicy: common.CronPolicy,
				}),
			},
		},
	}
```

**Review by cheyang — CHANGES_REQUESTED** [review]

File: cmd/juicefs/app/juicefs.go [hunk 5]
```diff
@@ -18,6 +18,7 @@ package app
 
 import (
 	"os"
+	metricsserver "sigs.k8s.io/controller-runtime/pkg/metrics/server"
```
Comment (cheyang) [inline cmd/juicefs/app/juicefs.go:21]: Could you please move 'metricsserver "[sigs.k8s.io/controller-runtime/pkg/metrics/server](http://sigs.k8s.io/controller-runtime/pkg/metrics/server)"' near '[sigs.k8s.io/controller-runtime/pkg/client](http://sigs.k8s.io/controller-runtime/pkg/client)'?

**Review by cheyang — CHANGES_REQUESTED** [review]

File: pkg/ddc/base/portallocator/bitmap_allocator.go [hunk 6]
```diff
@@ -36,7 +36,7 @@ func (b *BitMapAllocator) needResetReservedPorts() bool {
 }
 
 func newBitMapAllocator(pr *net.PortRange, log logr.Logger) (BatchAllocatorInterface, error) {
-	alloc, err := portallocator.New(*pr, func(max int, rangeSpec string) (allocator.Interface, error) {
+	alloc, err := portallocator.New(*pr, func(max int, rangeSpec string, offset int) (allocator.Interface, error) {
```
Comment (cheyang) [inline pkg/ddc/base/portallocator/bitmap_allocator.go:39]: Please add TODO for passing `offset` value. Thanks.

File: cmd/dataset/app/dataset.go (hunk 4 above)
Comment (xliuqq) [reply cmd/dataset/app/dataset.go:249]: no need to, as the `Scheme` field in `ctrl.Options`  will used to create the default Client and Cache.

File: cmd/juicefs/app/juicefs.go (hunk 5 above)
Comment (xliuqq) [reply cmd/juicefs/app/juicefs.go:21]: fixed

File: pkg/ddc/base/portallocator/bitmap_allocator.go (hunk 6 above)
Comment (xliuqq) [reply pkg/ddc/base/portallocator/bitmap_allocator.go:40]: fixed

File: cmd/juicefs/app/juicefs.go (hunk 3 above)
Comment (xliuqq) [reply cmd/juicefs/app/juicefs.go]: fixed

File: cmd/dataset/app/dataset.go (hunk 4 above)
Comment (cheyang) [reply cmd/dataset/app/dataset.go:249]: How about setting scheme in cache option to keep consistency with previous version? My thought is to make it easy for maintenance.  WDYT？

File: pkg/ddc/alluxio/node_test.go [hunk 7]
```diff
@@ -169,6 +177,10 @@ func TestSyncScheduleInfoToCacheNodes(t *testing.T) {
 				name:      "hbase-a",
 				namespace: "big-data",
 				worker: &appsv1.StatefulSet{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "StatefulSet",
+						APIVersion: "apps/v1",
+					},
```
Comment (TrafalgarZZZ) [inline pkg/ddc/alluxio/node_test.go:183]: Why is this necessary in new version of controller-runtime?

Comment (xliuqq) [reply pkg/ddc/alluxio/node_test.go:183]: It seems related to https://github.com/kubernetes-sigs/controller-runtime/pull/2633. If removed, test cases will fail.

**Review by cheyang — APPROVED** [review]
/lgtm
/approve

**Comment by cheyang** [comment]
@xliuqq Thanks for your great contribution!
